File: Medium/3. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.java
/*
LeetCode Medium: 3. Longest Substring Without Repeating Characters
https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

Check LongestSubstringWithoutRepeatingCharacters.cpp for brief solution explanation.
*/
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int N = s.length();
        if(N==0) return 0;
        int l = 0, r = 0, ans = 0;
        HashMap<Character, Integer>freq = new HashMap<Character, Integer>();
        while(l<N){
            while(r<N&&freq.getOrDefault(s.charAt(r), 0)==0){
                freq.put(s.charAt(r), freq.getOrDefault(s.charAt(r), 0)+1);
                r++;
            }
            ans = Math.max(ans, r-l);
            freq.put(s.charAt(l), 0);
            l++;
        }
        ans = Math.max(ans, r-l);
        return ans;
    }
}

File: Medium/3. Longest Substring Without Repeating Characters/LongestSubstringWithoutRepeatingCharacters.py
'''
LeetCode Medium: 3. Longest Substring Without Repeating Characters
https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

Check LongestSubstringWithoutRepeatingCharacters.cpp for brief solution explanation.
'''
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        N = len(s)
        if(N==0):
            return 0
        l, r, ans, = 0, 0, 0
        freq = defaultdict(int)
        while l<N:
            while r<N and freq[s[r]]==0:
                freq[s[r]]+=1
                r+=1
            ans = max(ans, r-l)
            freq[s[l]]-=1
            l+=1
        ans = max(ans, r-l)
        return ans
